pkg/api/dirs.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"mime"
+	"net/http"
+	"path/filepath"
+	"strings"
+
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/manifest/jsonmanifest"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// dirUploadInfo contains the data for a directory to be uploaded
+type dirUploadInfo struct {
+	toEncrypt bool
+	reader    io.ReadCloser
+}
+
+// dirUploadHandler uploads a directory supplied as a tar in an HTTP request
+func (s *server) dirUploadHandler(w http.ResponseWriter, r *http.Request) {
+	dirInfo, err := getDirHTTPInfo(r)
+	if err != nil {
+		s.Logger.Errorf("dir upload, get dir info: %v", err)
+		jsonhttp.BadRequest(w, "could not extract dir info from request")
+		return
+	}
+
+	reference, err := storeDir(r.Context(), dirInfo, s.Storer, s.Logger)
+	if err != nil {
+		s.Logger.Errorf("dir upload, store dir: %v", err)
+		jsonhttp.InternalServerError(w, "could not store dir")
+		return
+	}
+
+	jsonhttp.OK(w, fileUploadResponse{
+		Reference: reference,
+	})
+}
+
+// getDirHTTPInfo extracts data for a directory to be uploaded from an HTTP request
+func getDirHTTPInfo(r *http.Request) (*dirUploadInfo, error) {
+	if r.Body == http.NoBody {
+		return &dirUploadInfo{}, errors.New("request has no body")
+	}
+	toEncrypt := strings.ToLower(r.Header.Get(EncryptHeader)) == "true"
+	return &dirUploadInfo{
+		reader:    r.Body,
+		toEncrypt: toEncrypt,
+	}, nil
+}
+
+// storeDir stores all files recursively contained in the directory given as a tar
+// it returns the hash for the uploaded manifest corresponding to the uploaded dir
+func storeDir(ctx context.Context, dirInfo *dirUploadInfo, s storage.Storer, logger logging.Logger) (swarm.Address, error) {
+	dirManifest := jsonmanifest.NewManifest()
+
+	// set up HTTP body reader
+	tarReader := tar.NewReader(dirInfo.reader)
+	defer dirInfo.reader.Close()
+
+	// iterate through the files in the supplied tar
+	for {
+		fileHeader, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return swarm.ZeroAddress, fmt.Errorf("read tar stream error: %v", err)
+		}
+
+		filePath := fileHeader.Name
+
+		// only store regular files
+		if !fileHeader.FileInfo().Mode().IsRegular() {
+			logger.Warningf("skipping file upload for %s as it is not a regular file", filePath)
```

Comment:
this can go into `Debug` or even `Trace`

Reply:
i disagree, for me this is an arguably unexpected piece of information about business logic which is worth warning about.

Reply:
...